[tt-train] Add support for importing Llama 3 model weights

jaykru-tt opened the pull request:
### Problem description
We need to import weights from pretrained Llama 3 models in order to finetune them with tt-train.

### What's changed
- New model yaml configs for TinyLlama and Llama 3.2 1B
- Various changes to nanogpt/main.cpp to support dumping and loading model msgpacks.
- Formatters for xtensors and xtensor shapes. Found this very useful for debugging. Not a ton of code, so I think we can merge it with this PR.
- Adds support for fix intermediate dims to the tt-train Llama implementation. This was required to support Llama 3.2 1B.
- Adds support for NTK-aware RoPE scaling for extended context lengths. This is also used by Llama 3.2 1B.
- Adds a test for the TinyLlama tokenizer; not sure whether I will end up putting the tokenizer json in the repo, so we may want to nix this.
- Adds a Llama 3 weight loading script + documentation and some minor tweaks to the existing GPT2 weight loading script.

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

jaykru-tt committed:
tinyllama import working

dmakoviichuk-tt commented on tt-train/sources/ttml/ops/rope_op.cpp:
```diff
@@ -72,15 +68,52 @@ void validate_rope_input_and_params(const autograd::TensorPtr& input, const Rota
 }
 
+template <typename E>
+E apply_ntk_aware_scaling(const E& freqs, const NTKAwareScalingParams& scaling_params) {
```
d owe really need template here? can we just pass xtensor?

jaykru-tt replied:
Needs a template to be able to handle lazy xtensor expressions as well as realized xarrays.

jaykru-tt replied:
We should add this to core::from_xtensor btw. Will help us avoid having to force everything to an xarray all over the place 😄

rfurko-tt commented on tt-train/sources/ttml/models/llama.cpp:
```diff
@@ -55,11 +59,30 @@ Llama::Llama(const LlamaConfig& config) {
 
-    m_rope_params = ops::build_rope_params(max_sequence_length, embedding_dim / num_heads);
+    // Create NTK-aware scaling params if they are set
+    ops::NTKAwareScalingParams ntk_params;
```
what is ntk?

rfurko-tt commented on tt-train/sources/ttml/models/llama.cpp:
```diff
@@ -55,11 +59,30 @@ Llama::Llama(const LlamaConfig& config) {
+        fmt::print("        Low freq factor: {}\n", config.low_freq_factor);
+    }
+
+    m_rope_params = ops::build_rope_params(
```
could you please also make this change for distributed llama model?

rfurko-tt commented on tt-train/tools/weight-utils/llama_export.py:
```diff
@@ -0,0 +1,340 @@
+            "attention_norm/gamma/value/data": "input_layernorm.weight",
+            "mlp_norm/gamma/value/data": "post_attention_layernorm.weight",
+            "attention/q_linear/weight/value/data": "self_attn.q_proj.weight",
+            # "attention/q_linear/bias/value/data": None,
```
remove comments

rfurko-tt commented on tt-train/tools/weight-utils/llama_export.py:
```diff
@@ -0,0 +1,340 @@
+
+msgpack_numpy.patch()
+
+parser = argparse.ArgumentParser()
```
move parser in main function

rfurko-tt commented on tt-train/tools/weight-utils/llama_export.py:
```diff
@@ -0,0 +1,340 @@
+            # "mlp/w3/bias/value/data": None,
+        }
+
+        if block_key_map.get(rest):
```
is it the same as `if rest in block_key_map`?

rfurko-tt commented on tt-train/tools/weight-utils/llama_export.py:
```diff
@@ -0,0 +1,340 @@
+# /// script
+# requires-python = ">=3.11"
+# dependencies = [
```
what is this? is it working script or just helpful information? 

if needed, you can add `requirements.txt` for people being able to setup environment from scratch

rfurko-tt commented on tt-train/tools/weight-utils/llama_export.py:
```diff
@@ -0,0 +1,340 @@
+args = parser.parse_args()
+
+
+@ipdb.iex
```
what is ipdb?

rfurko-tt replied:
isn't it debug code?

jaykru-tt replied:
this decorator launches into a debugger post-mortem on an exception; since this script will only be used by devs I think it's a nice thing to leave in place.

rfurko-tt commented on tt-train/tools/weight-utils/README.md:
```diff
@@ -0,0 +1,67 @@
+
+This folder contains a collection of scripts for exporting the weights of
+various Huggingface models into a format suitable for use in tt-train. It is
+recommended to run these scripts with [`uv`](https://docs.astral.sh/uv/) which
```
do we need to start using new python package? :D

jaykru-tt replied:
yes, it's so much better

jaykru-tt committed:
revert generate changes, rope test, pub/private

jaykru-tt committed:
revert deps, msgpack_file, remove llama_test and test gen stuff

jaykru-tt committed:
rename llama export script

jaykru-tt committed:
weight-utils dir + readme

jaykru-tt committed:
fix xtensor includes

jaykru-tt committed:
some tokenizer dump tweaks

jaykru-tt committed:
revert some unneeded changes

jaykru-tt committed:
move eval back to before optimizer setup

jaykru-tt committed:
explicit init in scaling params

jaykru-tt committed:
add copyright

jaykru-tt committed:
fix rope ntk scaling issue

Revert "try reverting ntk"

This reverts commit dde91265257511d763129b1d85b75a4fcdd5a46d.

add support for just dumping the tokenizer to llama export script

fix ntk issue with rope op

jaykru-tt committed:
:p

jaykru-tt committed:
fix dumping code for gpt2

jaykru-tt committed:
make gpt2 tokenizer dumping more robust:

point hf-transformers to cache in a directory under /tmp, load the
gpt2tokenizer there then copy it to the desired path; for some reason
this is the only way I managed to get a working gpt2s tokenizer loaded
:(

jaykru-tt committed:
Address Roman's comments

jaykru-tt committed:
fixes

jaykru-tt committed:
roman comments

jaykru-tt committed:
update readme about venv change

jaykru-tt committed:
NTKAware -> Rope

simpler naming, less likely to get terms of art wrong

rfurko-tt commented on tt-train/tools/weight-utils/requirements.txt:
```diff
@@ -0,0 +1,6 @@
+ipdb
```
thanks!

rfurko-tt commented on tt-train/tools/weight-utils/gpt2_export.py:
```diff
@@ -1,3 +1,4 @@
+#!/usr/bin/env python3
```
do we need this line?

rfurko-tt commented on tt-train/sources/ttml/modules/llama_block.cpp:
```diff
@@ -5,16 +5,20 @@
+        hidden_size = *intermediate_dim;
+    } else {
+        const uint32_t unrounded_size = static_cast<uint32_t>(static_cast<float>(4 * embedding_size) * (2.0F / 3.0F));
+        hidden_size = ((unrounded_size + multiple_of - 1) / multiple_of) * multiple_of;
```
1U

rfurko-tt reviewed (commented):
Looks good overall, please update DistributedLLaMaMLP as well, thanks in advance!

jaykru-tt committed:
Merge branch 'main' into jkruer/llama_import

jaykru-tt committed:
intermediate dim support for dist llama block

jaykru-tt committed:
roman comments

jaykru-tt commented:
Looks like I covered everything. Merging now 🚀
